internal: remove transportMonitor, replace with callbacks

jadekler opened the pull request:
This removes a goroutine (in which transportMonitor was running) and the backoff and addrs loops. Instead we use callbacks to handle the deadline, goaway, and close logic that transport monitor handled before, and we descend into recursive createTransport calls instead of loops.

As a result, backoffIdx and addrIdx now live on addrConn. Also, transport.Close now blocks until the addrConn that initiated it is connected - this is because t.Close calls onClose, whose last step is to begin reconnecting. In the event of a shutdown, all reconnect logic ends and t.Close will return.

jadekler commented on clientconn.go:
```diff
@@ -1515,6 +1517,20 @@ func (ac *addrConn) tearDown(err error) {
 	}
+
+	// TODO(deklerk) is this bad?
+	transports := map[int32]transport.ClientTransport{}
```
Naive way to get around needing the transports outside the lock (GracefulClose takes the lock), but not wanting to cause data races.

Is there a better way to do this?

jadekler commented on stream.go:
```diff
@@ -261,6 +261,11 @@ func newClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 
+		// TODO(deklerk) what??
+		if t == nil {
+			continue
```
I couldn't figure out why this is happening, but sometimes I would get nil, true, nil which blows up a little below when we call t.NewStream. This conditional fixes the problem, but I thought I'd bring it up. I suspect it's a race between the transport closing - and getting deleted - and getTransport looking for, finding, and returning it.

If so, is this fix appropriate or should I make getTransport and transport deletion synchronized? Alternatively, should I document getTransport's race with closing transport? (again, all this predicated on my guess above)

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1178,217 @@ func (ac *addrConn) resetTransport() error {
-		}
-		if connected {
-			return nil
+		if t == nil {
```
I'm not sure this solves the problem appropriately. AFAICT there's a race between the reader goroutine closing and onDeadline. This puts a plaster over the problem but I think there's a still a chance t.Close just below could happen incorrectly. Is that accurate? Is there a better way to handle this? Do I need to do more synchronization?

dfawley commented:
Yes, that's right.  It's only failing in 1.10 because we don't run the vet/lint/etc checks other builds.

jadekler commented:
@dfawley Is there a way to see what cause it to fail? Or, repro locally? `gofmt -s -d -l .` returns nothing for me, but not sure if I missed a step.

dfawley commented:
What version is your local go?

...Now I'm even more confused.  The output shows it grepping for "unsafe", but if I look in your branch, your vet.sh doesn't have the change to do that yet.

I think we should remove the "-l" parameter from the command-line so we can see the full diffs.  You can do that in your branch, anyway, to get travis to output it.

dfawley commented:
> I'm running go1.10

Missed that.  Are you not running 1.10.1?  That's what travis used.

jadekler commented:
> Missed that. Are you not running 1.10.1? That's what travis used.

Updated and rebased with master, still not seeing anything with `vet.sh`. I took your advice and removed `-l`; we'll see what travis prints.

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+	newTrID := atomic.AddInt32(&ac.transportIdx, 1)
+
+	onGoAway := func() {
+		shutdownLock.Lock()
```
timer. I'll tighten it up.

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
 		ac.ready = nil
```
Done

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+		if err != nil {
+			return err
 		}
+		return ac.resetTransport()
```
Done

jadekler commented:
@MakMukhi Updated from recursion to loop

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+		case <-connectCtx.Done():
+			// Didn't receive server preface, must kill this new transport now.
+			grpclog.Warningf("grpc: addrConn.createTransport failed to receive server preface before deadline.")
+			newTr.Close()
```
Done

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
-			case <-ac.ctx.Done():
-			}
+
+		err = ac.resetTransport()
```
I believe that is the current behavior, no?

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
-		newTr, err := transport.NewClientTransport(connectCtx, ac.cc.ctx, target, copts, onPrefaceReceipt)
+
+		ac.mu.Lock()
+		delete(ac.transports, newTrID)
```
Good call. Done

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+	shutdownLock := &sync.Mutex{}
+	var timer *time.Timer
+	ac.mu.Lock()
+	if !ac.connectDeadline.IsZero() {
```
Ah, never. Removed conditional

jadekler commented:
Just realized I was pushing the wrong branch heh. The updates I promised above are now actually here :)

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+		shutdownLock.Lock()
+		defer shutdownLock.Unlock()
+
+		if timer != nil {
```
You're right. Done.

jadekler commented on clientconn.go:
```diff
@@ -1174,152 +1179,217 @@ func (ac *addrConn) resetTransport() error {
+		if t == nil {
+			return
 		}
+		t.Close()
```
Just so I'm clear, since the diff is a bit hard to read; you're suggesting that we remove onDeadline and move the onDeadline logic into the transport reader goroutine by way of passing connectDeadline down to the transport?

jadekler commented on clientconn.go:
```diff
@@ -1,3 +1,11 @@
+/**
```
Whoops, I should and will remove this

jadekler commented:
@MakMukhi PTAL.

- Refactor to use reader goroutine.
- Rename back to resetTransport.
- Various review comment fixes and cleanups.
- Found a home for resolving (in nextAddr).
- Rebased with master.

jadekler commented:
Now that location differs, I'm closing this and re-opening in a new, clean PR.
